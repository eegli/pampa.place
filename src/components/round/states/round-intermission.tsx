import {startOrResumeRound} from '@/redux/game';
import {shouldRequestNewSV} from '@/redux/game/selectors';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import {resetSelectedPosition} from '@/redux/position';
import {getRandomStreetView} from '@/redux/position/thunks';
import {Button, Divider, Stack, Typography} from '@mui/material';
import {useLayoutEffect} from 'react';
import {SlimContainer} from '../../../styles/containers';

export const RoundIntermission = () => {
  const dispatch = useAppDispatch();

  const isLoadingStreetView = useAppSelector(({position}) => position.loading);
  const currentPlayer = useAppSelector(({game}) => game.players[0]);
  const currentRound = useAppSelector(({game}) => game.rounds.current);
  const totalRounds = useAppSelector(({game}) => game.rounds.total);
  const shouldGetNewSV = useAppSelector(shouldRequestNewSV);

  useLayoutEffect(() => {
    if (shouldGetNewSV) {
      dispatch(getRandomStreetView());
    }
  }, [dispatch]);

  function handleClick() {
    dispatch(resetSelectedPosition());
    dispatch(startOrResumeRound());
  }

  return (
    <SlimContainer height="100%" id="c-round-intermission">
      <Stack direction="column" alignItems="center" spacing={3} margin="auto">
        <Typography variant="h4" align="center">
          {currentPlayer}, it&apos;s your turn!
        </Typography>
        <Divider orientation="horizontal" flexItem />
        <Typography
          variant="h6"
          sx={{
            color: 'text.secondary',
          }}
        >
          Round {currentRound}/{totalRounds}
        </Typography>

        <Button
          onClick={handleClick}
          variant="contained"
          color="primary"
          disabled={isLoadingStreetView}
        >
          {isLoadingStreetView
            ? 'Getting a random Street View...'
            : 'Start Round'}
        </Button>
      </Stack>
    </SlimContainer>
  );
};
