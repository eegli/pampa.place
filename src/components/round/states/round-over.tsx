import {endRound} from '@/redux/game';
import {getRoundScores, isFinished} from '@/redux/game/selectors';
import {useAppDispatch, useAppSelector} from '@/redux/hooks';
import {formatDist} from '@/utils/misc';
import {
  Box,
  Button,
  Paper,
  Stack,
  Tab,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  Tabs,
  Typography,
} from '@mui/material';
import {SyntheticEvent, useState} from 'react';
import {SlimContainer} from '../../../styles/containers';
import {GoogleMap} from '../../google/map';
import {GoogleStreetView} from '../../google/street-view';
import {TabPanel} from './tabs-panel';

export const RoundOverSummary = () => {
  const [selectedPanel, setSelectedPanel] = useState(0);

  const handleChange = (event: SyntheticEvent, newValue: number) => {
    setSelectedPanel(newValue);
  };
  const dispatch = useAppDispatch();
  const activeMapId = useAppSelector(({game}) => game.mapId);
  const scores = useAppSelector(getRoundScores);
  const currentRound = useAppSelector(({game}) => game.rounds.current);
  const initialPosition = useAppSelector(
    ({position}) => position.initialPosition
  );

  const isGameFinished = useAppSelector(isFinished);

  const panoDescription = useAppSelector(
    ({position}) => position.panoDescription
  );

  function handleClick() {
    dispatch(endRound());
  }

  const text = isGameFinished
    ? 'See results!'
    : `Continue with round ${currentRound + 1}`;

  return (
    <>
      <Box py={2}>
        <Tabs value={selectedPanel} onChange={handleChange} centered>
          <Tab label="Result" />
          <Tab label="Map" />
          <Tab label="Street View" />
          <Tab label="Info" />
        </Tabs>
      </Box>

      <TabPanel
        selected={selectedPanel}
        index={0}
        display="flex"
        justifyContent="center"
        /*   alignItems="center" */
      >
        <SlimContainer id="round-end-table">
          <Stack
            direction="column"
            alignItems="center"
            spacing={3}
            width="100%"
            height="100%"
          >
            <Typography variant="h5" alignSelf="flex-start">
              Round {currentRound} is over!
            </Typography>

            <Table size="small">
              <TableHead>
                <TableRow>
                  <TableCell>Name</TableCell>
                  <TableCell align="right">Distance</TableCell>
                  <TableCell align="right">Score</TableCell>
                </TableRow>
              </TableHead>
              <TableBody>
                {scores.map((p, idx) => (
                  <TableRow key={p.name}>
                    <TableCell component="th" scope="row">
                      {p.name}
                    </TableCell>
                    <TableCell align="right">{formatDist(p.dist)}</TableCell>
                    <TableCell align="right">{p.score}</TableCell>
                  </TableRow>
                ))}
              </TableBody>
            </Table>
            <Button onClick={handleClick} variant="contained" color="primary">
              {text}
            </Button>
          </Stack>
        </SlimContainer>
      </TabPanel>

      <TabPanel selected={selectedPanel} index={1}>
        <Box height="100%" id="round-end-map">
          {initialPosition && (
            <GoogleMap
              mode="result"
              mapId={activeMapId}
              results={scores}
              initialPosition={initialPosition}
            />
          )}
        </Box>
      </TabPanel>

      <TabPanel selected={selectedPanel} index={2}>
        <Box height="100%" id="round-end-sv">
          <GoogleStreetView staticPos />
        </Box>
      </TabPanel>

      <TabPanel selected={selectedPanel} index={3}>
        <SlimContainer id="round-end-loc-info">
          <Typography variant="h5" alignSelf="flex-start" mb={2}>
            Location info
          </Typography>

          <Paper sx={{p: 2, mb: 1}} elevation={1}>
            Panorama description: {panoDescription}
          </Paper>
          <Paper sx={{p: 2}}>
            Coordinates (lat/lng): {initialPosition?.lat},{' '}
            {initialPosition?.lng}
          </Paper>
        </SlimContainer>
      </TabPanel>
    </>
  );
};
