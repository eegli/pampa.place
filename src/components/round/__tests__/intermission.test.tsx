import {createMockState, createMockStore, render} from '@/tests/test-utils';
import {RoundIntermission} from '../states/intermission';

const mockPush = jest.fn();

jest.mock('next/router', () => ({
  useRouter() {
    return {
      push: mockPush,
    };
  },
}));

const mockRes = {
  data: {
    copyright: 'copyright',
    imageDate: 'sv image',
    links: [],
    location: {
      description: 'mock panorama description',
      pano: 'mock panorama id',
      shortDescription: null,
      latLng: {
        lat() {
          return 1;
        },
        lng() {
          return 2;
        },
      },
    },
  },
} as unknown as google.maps.StreetViewResponse;

jest
  .spyOn(google.maps.StreetViewService.prototype, 'getPanorama')
  .mockResolvedValue(mockRes);

describe('Round intermission', () => {
  it('fetches new random location at the start of round', async () => {
    const state = createMockState({
      game: {
        players: ['player 1', 'player 2'],
        rounds: {
          total: 3,
          current: 1,
          // If round progress is 0, get a new location
          progress: 0,
        },
      },
      position: {
        initialPosition: null,
      },
    });

    const store = createMockStore(state);
    render(<RoundIntermission />, store);
    // Somehow this is necessary for the store to update.
    await new Promise(r => setTimeout(r, 0));
    expect(store.getState().position).toMatchSnapshot('new random location');
  });
});
