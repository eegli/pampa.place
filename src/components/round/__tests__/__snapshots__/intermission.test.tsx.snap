// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Round intermission fetches new random location at the start of round: new random location 1`] = `
Object {
  "error": null,
  "initialPosition": Object {
    "lat": 1,
    "lng": 2,
  },
  "loading": false,
  "panoDescription": "mock panorama description",
  "panoId": "mock panorama id",
  "selectedPosition": null,
}
`;
